addons/payment_paypal/controllers/main.py

```diff
@@ -24,52 +26,104 @@ def paypal_dpn(self, **data):
         The route also allows the GET method in case the user clicks on "go back to merchant site".
         """
         _logger.info("beginning DPN with post data:\n%s", pprint.pformat(data))
-        try:
-            self._validate_data_authenticity(**data)
-        except ValidationError:
-            pass  # The transaction has been moved to state 'error'. Redirect to /payment/status.
+        if not data:  # The customer has cancelled the payment
+            pass  # Redirect them to the status page to browse the draft transaction
         else:
-            if data:
-                request.env['payment.transaction'].sudo()._handle_feedback_data('paypal', data)
+            try:
+                notification_data = self._validate_pdt_data_authenticity(**data)
+            except ValidationError:
+                pass  # The data authenticity could not be verified, discard the notification
             else:
-                pass  # The customer has cancelled the payment, don't do anything
+                request.env['payment.transaction'].sudo()._handle_feedback_data(
+                    'paypal', notification_data
+                )
+
         return request.redirect('/payment/status')
 
+    def _validate_pdt_data_authenticity(self, **data):
+        """ Validate the authenticity of PDT data and return the retrieved notification data.
+
+        The validation is done in four steps:
+          - 1: Make a POST request to Paypal with the `tx`, the GET param received with the PDT,
+               data, and the two other required params `cmd` and `at`.
+          - 2: PayPal sends back a response text starting with either 'SUCCESS' or 'FAIL'. If the
+               validation was a success, the notification data are appended to the response text as
+               a string formatted as follows: 'SUCCESS\nparam1=value1\nparam2=value2\n...'
+          - 3: Extract the notification data and process these instead of the PDT data.
+          - 4: Return an empty HTTP 200 response (done at the end of the route controller).
+        See https://developer.paypal.com/docs/api-basics/notifications/payment-data-transfer/.
+
+        :param dict data: The data whose authenticity to check
+        :return: The retrieved notification data
+        :raise: ValidationError if the authenticity could not be verified
+        """
+        def _parse_pdt_validation_response(_response_content):
+            """ Parse the validation response and return the parsed notification data.
+
+            :param str _response_content: The PDT validation request response
+            :return: The parsed notification data
+            :rtype: dict
+            """
+            _response_items = _response_content.rstrip('\n').split('\n')
```

Comment:
> Why is everything prefixed with `_`?

It's a convention to easily distinguish variables defined in the nested function's scope and to avoid shadowing a variable from the calling method's scope.

> Also `response_content.splitlines()` would do the entire job on its own without having to deal with weird trimming.

Indeed. I didn't realize that `splitlines` takes care of trailing `\n`.